ibis/expr/types/core.py

```diff
@@ -217,23 +217,16 @@ def _find_backends(self) -> tuple[list[BaseBackend], bool]:
         list[BaseBackend]
             A list of the backends found.
         """
-        import ibis.expr.operations as ops
-        from ibis.backends.base import BaseBackend
-
-        def finder(node):
-            # BaseBackend objects are not operation instances, so they don't
-            # get traversed, this is why we need to select backends out from
-            # the node's arguments
-            backends = [arg for arg in node.args if isinstance(arg, BaseBackend)]
-            return g.proceed, (backends, isinstance(node, ops.UnboundTable))
-
-        all_backends = []
-        any_unbound = False
-        for backends, has_unbound in g.traverse(finder, self.op()):
-            all_backends += backends
-            any_unbound |= has_unbound
-
-        return list(toolz.unique(all_backends)), any_unbound
+        backends = set()
+        has_unbound = False
+        node_types = (ops.DatabaseTable, ops.SQLQueryResult, ops.UnboundTable)
```

Comment:
Perhaps it would be even better to define a base node which specifies the `.source` property, like

```python
class PhyisicalOp(ops.Node):
    source: Any
```

Though I'm not insisting to this for now.

Reply:
Yeah, perhaps `UnboundTable` is no longer a type even, and it's just a `PhysicalTable` (or `PhysicalOp`) with `source = None`.